mlflow/tracing/provider.py

```diff
@@ -36,3 +48,96 @@ def _setup_tracer_provider(client: Optional[TraceClient] = None):
     tracer_provider = TracerProvider()
     tracer_provider.add_span_processor(SimpleSpanProcessor(exporter))
     trace.set_tracer_provider(tracer_provider)
+
+
+def create_trace_info(
+    otel_span: trace.Span,
+    experiment_id: Optional[str] = None,
+    request_metadata: Optional[dict] = None,
+    tags: Optional[dict] = None,
+) -> TraceInfo:
+    """
+    Create a new TraceInfo object based on the OpenTelemetry span and the given metadata.
+
+    The TraceInfo generation logic depends on the environment where the trace is created,
+    and this function is responsible for encapsulating the differences.
+
+    Args:
+        otel_span: The OpenTelemetry span object.
+        experiment_id: The experiment ID for the trace.
+        request_metadata: The metadata of the request.
+        tags: The tags of the trace.
+    """
+    from mlflow.tracking.fluent import _get_experiment_id
+
+    # If the environment is Databricks, the trace should be logged in the tracking server.
+    # The initial TraceInfo entry is created by calling StartTrace backend API.
+    if is_in_databricks_runtime():
+        from mlflow.tracking.client import MlflowClient
+
+        try:
```

Comment:
Seems a bit unfortunate / complex that we have the `is_in_databricks_runtime()` check in two places: tracing client and tracing provider. Is there a way we can refactor such that we only have to check the environment in one place?

Reply:
Hmm true... not 100% sure if there is very clean way, but will give some thought on it.

Reply:
While I explored a few more options, I just remembered that the span processor interface has `on_start()` handler ([ref](https://github.com/open-telemetry/opentelemetry-python/blob/47d5ad7aae5aef31238ca66e55dc550b307c7b35/opentelemetry-sdk/src/opentelemetry/sdk/trace/export/__init__.py#L103)). This seems to be a good place to put those `StartTrace()` call and `TraceInfo` initialization. This allows us to encapsulate the above mentioned two factory logic into single place, inside `_setup_tracer_provider()` method.

The idea is to remove `TraceClient`, and instead defines `SpanProcessor` (with existing `Exporter`). The span processor takes responsibility of creating trace info and adjusting it on span completion, while the exporter accounts for the logging (which is similar to the client's responsibility before).

The trace lifecycle will be like
1.  When the very first tracing starts, `_setup_tracer_provider()` is called to set up tracer and other components. Here `MlflowSpanProcessor` and `MlflowExporter` are initialized in Databricks notebook, which can vary across environment like model serving.
2. When a span starts, `on_start()` handler of the span processor is called and it invokes backend `StartTrace` API. The returned `TraceInfo` is registered to the in-memory trace manager. The `LiveSpan` wrapper is created by the same way as now.
3. When the span ends, `on_end()` handler of the span processor is called. It updates the stored `TraceInfo` with the latest span fields, as well as a few post processing e.g. span name deduplication. All trace mutation should be done at this phase.
4. The span processor calls `exporter.export()` that logs the trace to the backend. Here the exporter only focuses on writing and doesn't mutate the trace data at all.

At high-level, it looks like we shift the components from Exporter&Clients to Processor&Exporter, but the crucial difference is that the processor has `on_start()` handler too. This allows us to modularize both start and end logic together.

I've created a POC implementation here: https://github.com/B-Step62/mlflow/tree/refactor-exporter, including [MlflowSpanProcessor](https://github.com/B-Step62/mlflow/blob/refactor-exporter/mlflow/tracing/processor/mlflow.py) and updated [MlflowSpanExporter](https://github.com/B-Step62/mlflow/blob/refactor-exporter/mlflow/tracing/export/mlflow.py). Overall I feel this makes the architecture more clearer and also aligns with the OTel's design. Please let me know your thoughts on this idea!

Reply:
@B-Step62 This looks great - let's do it! Thanks for thinking about this and identifying a clean solution!